### ISSUE TITLE
Use SRI to Detect More Libraries

### ISSUE THREAD

**Issue by chris-barry** [Synzvato/decentraleyes#26]
Mozilla recently put in [SRI](https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity) support into Firefox. If a script tag has a SRI attached, and it matches the hash of a known, then it can be assumed there is no need to download the resource.

**Comment by austinhartzheim** [comment]
@chris-barry Just to clarify, are you suggesting that this be used to detect common libraries that are self-hosted?

If so, I think it is a good idea overall. However, some users might want to disable this functionality as it is a fingerprintable trait. That is, the fact that this add-on is installed could be detected by a web page you visit due to the behavior you describe.

**Comment by chris-barry** [comment]
Yes, that's exactly what I am suggesting. And I completely agree about the fingerprintability. It should be opt-in, maybe under an "aggressive" mode, or similar.

**Comment by Synzvato** [comment]
@chris-barry @austinhartzheim

Currently, Decentraleyes doesn't get along well with ```integrity``` and ```crossorigin```. In fact, all pages that implement these new attributes will not be able to load any resources unless they're whitelisted.

The issue, as described in #16, is quite hard to tackle as it seems that there's no way to tell if these attributes are present in the DOM when looking at an intercepted HTTP request. Also, the corresponding page is still mostly unparsed when the requests are sent out. So we'll need to find another way.

Once the related issue is solved (there has to be a way to detect the presence of these attributes on request interception), we can indeed use the resulting information to power this feature. Thanks for the suggestion! Any further suggestions would, of course, be greatly appreciated.

**Comment by mozfreddyb** [comment]
This is a really neat idea! But allow me to briefly explain a security problem that comes with using an SRI hash for caching:

Assume somone has a very secure website, `supersecure.example`, protected by [Content Security Policy (CSP)](https://developer.mozilla.org/en-US/docs/Web/Security/CSP). This high-security policy disallows scripts to be loaded from any domain other than `supersecure.example`. Given the defaults of CSP, there is no inline JavaScript on `supersecure.example` (e.g., `<script>…code here…</script>`), nor is it allowed.

Now, let's also assume there is a trivial XSS vulnerability at `supersecure.example/hello.php`, which will reflect whatever content you put in the name GET variable. The CSP prevents this bug from being exploitable (no new inline scripts, no remote script loads from somewhere other from `supersecure.example`.

Here come's a way to attack this with a cache poisoning attack:

An attacker can inject something that includes `<script src="supersecure.example/does-not-even-exist.js" integrity=foo>`, where integrity matches the hash of a known-vulnerable file. It is, for example, known that you can [bypass CSP and cause XSS with an outdated angular-js version](http://www.slideshare.net/x00mario/an-abusive-relationship-with-angularjs). The browser (extension) would happily include the old library, even though the file does not exist on `supersecure.example`. This bypasses the CSP restriction and makes the previously secure web page vulnerable.
I'm afraid there is no way to verify that the script was meant to be included and actually exists on said domain until you do the request (which, of course, you want to avoid).


*NB: Using SRI for caching is a neat idea and we're still looking at how to do so securely. Possibly via opt-in :/ Read more about that on the [w3c SRI issue tracker](https://github.com/w3c/webappsec-subresource-integrity/issues/22). If you have feedback, please provide your comments on the mailing list and not in the github issue)*

**Comment by stewie** [comment]
Please consider preserving an option to "fuggadabout SRI and (just) facilitate caching".

If a page author embeds a link to "jquery-2.0.1.min.js" (sez me) damn well better be the same file regardless whether intended to be retrieved from A) author's site, B) CDNx, or C) ajax.googleapis.com

To clarify: regardless the intended request URL specified by a page author, I would desire to serve MY permacached copy of "jquery-2.0.1.min.js". Hey, maybe I even want to hack it, intentionally removing or editing any (perceived as) objectionable portions of the script. BTW, given the option to do so, I would choose to NOT (retain redundant copies of and) serve myself minified versions of staple script libraries -- I would choose to match, instead, soley based on matching filename. Said differently, I would prefer to have multiple policy rules pointing various URLstrings to (one) same copy of  "jquery-2.0.1.min.js" (vs redundant copies retained in multiple mapped paths).

At reading earlier issues n comments here -- "plugins" and "subscriptions" and "whadabout the bluemoon possibility cache-stack-poisoning attacks" -- I worry that this extension is already becoming seriously over-engineered.

**Comment by stewie** [comment]
Will the "integrity" SRI attribute cause the browser to reject our permacached copy of a script when the "inject a comment denoting locally injected file" option is active?

**Comment by mozfreddyb** [comment]
If you modify the file, it will have a different hash. Even if it's just a comment. So, yes, it will be bocked.
